[5기 - 이중원] SpringBoot Part1 Weekly Mission 제출합니다.

shoeone96 opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
![VoucherSystem team - v2](https://github.com/prgrms-be-devcourse/springboot-basic/assets/85065626/941720c4-6390-4394-8673-b8b61cad99f6)

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
- console
    - **Console**
    - **Input**
    - **Output**
- consts
    - **ErrorMessage**
    - **Message**
- customer
    - controller
        - **CustomerController**
    - dto
        - **CustomerResponseDto**
    - model
        - **Customer**
        - **CustomerType**  
    - repository
        - **CustomerRepository**
        - **FileCustomerRepository**
    - service
        - **CustomerService**
- utils
    - **MenuType**
    - **Validation** 
- voucher
    - controller
        - **VoucherController**
    - dto
        - **VoucherRequestDto**
        - **VoucherResponseDto**
    - model
        - **Voucher**
        - **VoucherType**  
    - repository
        - **VoucherRepository**
        - **FileVoucherRepository**
        - **MemoryVoucherRepository**
    - service
        - **VoucherService**


## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->

1. Console을 Input과 Output의 구성 형태로 만들었는데 이런 식으로 만들었을 때 문제가 있을까요?

2. log를 메소드 요청의 흐름에 따라 기록하고 어느 로직에서 문제가 발생하는지 기록하고 싶은데 동시성 등을 고려해서 ThreadLocal을 사용해도 될까요? 쓰레드 로컬이 진짜 잘못 사용해서 메모리에 남으면 큰 영행을 미칠 수 있다고 들은 적이 있어서 고민했습니다.
(시간이 부족해서 구현까지는 못했습니다)

3. 동시성 처리를 위해 일종의 메시지 큐와 같이 큐에 담아서 요청에 대한 동시성을 처리하는 방법은 어떨지 궁금합니다.

shoeone96 committed:
init: 초기 설정

shoeone96 committed:
feat: VoucherApplication 생성
실제 서비스가 구동될 VoucherApplication class 생성 및 SpringOrderApplication에 CommandLinner로 연동

shoeone96 committed:
test: VoucherApplication test
VoucherApplication class 생성 이후 정상 작동 확인을 위한 테스트

yenzip committed:
feat: VoucherType 생성
바우처 타입 선택, 할인 금액 계산하는 메서드 생성

yenzip committed:
test: VoucherType 테스트
바우처 타입 유효성 테스트, 할인 금액 계산 테스트 수행

shoeone96 committed:
Merge pull request #1 from yenzip/main

[예림] VoucherType

shoeone96 committed:
feat: Voucher class 생성
도메인 객체 생성 및 계산 관련 메서드 작성

shoeone96 committed:
test: Voucher 도메인 관련 테스트
Voucher 생성 및 할인 적용 메서드 테스트

yenzip committed:
refactor: discountValue 원시 타입으로 변경
생성자 파라미터의 discountValue long 타입으로 통일

yenzip committed:
feat: VoucherRepository 생성
바우처 저장, 전체 조회, 번호로 조회, 초기화 메서드 생성

yenzip committed:
test: VoucherRepository test
바우처 저장, 전체 조회, 번호로 조회 성공/실패 메서드 테스트

shoeone96 committed:
Merge pull request #2 from yenzip/main

[예림] VoucherRepository

shoeone96 committed:
feat: voucher 저장을 위한 requestDto

voucher 객체 저장을 위해 사용자가 입력한 정보를 토대로 만든 객체

shoeone96 committed:
feat: Voucher 정보 반환 responseDto

Voucher 조회 시 필요한 정보를 반환하기 위해 만든 responseDto

shoeone96 committed:
feat: voucher 조회와 저장을 위한 서비스 로직

shoeone96 committed:
feat: request 및 response dto로 변환 시 필요한 생성자 및 메서드 추가

shoeone96 committed:
test: voucher 조회 및 저장 테스트

yenzip committed:
feat: MenuType enum 생성
프로그램 실행 메뉴 관련 객체

yenzip committed:
feat: Validation 생성
입력값 유효성 검사 수행

yenzip committed:
feat: Exception 생성
바우처 관련 custom exception

yenzip committed:
feat: Console 생성
입출력 관련 메서드

yenzip committed:
feat: 프로그램 실행 메뉴 추가
프로그램 종료, 바우처 생성, 조회 메서드 추가

yenzip committed:
feat: Log 추가
바우처 생성 확인을 위한 Log 추가

yenzip committed:
test: Application 구동 테스트 수행

shoeone96 committed:
Merge pull request #3 from yenzip/main

[예림] Console

shoeone96 committed:
feat: List 조회 메서드 추가

shoeone96 committed:
refactor: 출력 방식 변경

shoeone96 committed:
chore: EXIT 이후 종료가 안되는 부분 추후 수정 요

shoeone96 committed:
feat: list 출력 메서드 작성

yenzip committed:
fix: 메뉴 입력에서 메뉴 번호 입력으로 변경

yenzip committed:
test: @Disabled 처리
현재 무한 로딩으로 인한 임시 @Disabled 처리

yenzip committed:
test: 메뉴 입력 관련 Console 테스트
유효한 메뉴 번호 입력 성공 / 실패 테스트 수행

shoeone96 committed:
Merge pull request #4 from yenzip/main

[예림] Console 수정

shoeone96 committed:
feat: 애플리케이션 종료 기능 추가

exit 버튼을 누를 시 바우처 관리 애플리케이션을 종료할 수 있는 기능 완성

shoeone96 committed:
refactor: 입출력 콘솔 컴포넌트로 ApplicationContext에 등록

기존에 Console을 사용하던 클래스마다 인스턴스를 생성하던 것을 ApplicationContext에 등록하여 Singleton으로 사용하는 것으로 변경(메모리 낭비 감소)

shoeone96 committed:
refactor: VoucherApplication 책임 일부를 VoucherController에 이양

SRP원칙에 입각한 refactoring 구현

기존 VoucherApplication 역할
1. VoucherService 매핑
2. 사용자 입력 받아오기, entity 정보 출력

VoucherController 책임 이양
1. 사용자 입력 받아오기, entity 정보 출력 책임 이양

yenzip committed:
refactor: Console에서 Input과 Output 분리
SRP 원칙에 따라 기존 Console에서의 입출력을 Input과 Output을 사용해 구성으로 리팩토링

yenzip committed:
refactor: 메세지 분리
기존 Console에서 메세지를 Message와 ErrorMessage를 분리해 리팩토링 진행

shoeone96 committed:
Merge pull request #5 from yenzip/main

[예림] Console 및 Message

shoeone96 committed:
chore: 불필요한 테스트 케이스 삭제

shoeone96 committed:
test: 예외처리 완료로 예외가 발생하지 않는 테스트 @Disabled

shoeone96 committed:
feat: maximum value, minimum value 추가 및 VoucherType 구분 숫자 변경

FIXED: range: 100~10000
PERCENT: range: 3~50
voucherNum type: int -> String

shoeone96 committed:
test: MenuType 구분 숫자의 자료형 변환된 부분 test 코드 반영

shoeone96 committed:
feat: console로 받아오는 내용 예외처리

shoeone96 committed:
refactor: message 상수값 사용으로 통일

yenzip committed:
feat: yaml 프러퍼티 생성
default, dev, local 환경 설정

yenzip committed:
chore: application.properties 파일 삭제

yenzip committed:
feat: FileRepository 생성
파일로 관리되는 FileRepository 추가

yenzip committed:
chore: 사용하지 않는 clear 메서드 삭제

yenzip committed:
feat: FileRepository에 사용되는 ErrorMessage 추가

yenzip committed:
chore: 사용하지 않은 clear 메서드 삭제

yenzip committed:
test: VoucherService 테스트 순서 Order로 지정

yenzip committed:
feat: logback 설정해 에러를 파일로 기록

shoeone96 committed:
Merge pull request #6 from yenzip/main

[예림] FileRepository 추가 및 Logback 설정

shoeone96 committed:
refactor: domain 추가에 따른 폴더 구조 변경

shoeone96 committed:
chore: 안쓰는 import 문 삭제

shoeone96 committed:
refactor: 프로그램 종료 메세지 추가

shoeone96 committed:
feat: black list를 조회할 수 있는 repository 클래스 및 기능 추가

csv 파일 기반으로 조회 가능

shoeone96 committed:
feat: csv 파일에 CRUD를 위한 Customer domain entity

shoeone96 committed:
feat: blacklist 고객 조회를 위한 서비스 로직 생성

shoeone96 committed:
feat: 블랙리스트 조회 api controller

추후 고객 관련 api 추가 예정

shoeone96 committed:
feat: controller에서 black list 출력을 위한 dto, console 로직 생성

shoeone96 committed:
feat: black list에 회원이 존재하지 않을 경우 출력하는 메시지 추가

shoeone96 committed:
feat: voucherApplication에 blacklist 회원 조회 메뉴 추가

shoeone96 committed:
chore: 실제 customerList에서 black list 추출해서 조회하는 로직

shoeone96 committed:
chore: 파일 이름 변경

blackList -> customerList

yenzip committed:
feat: @Profile 어노테이션 적용

yenzip committed:
refactor: voucher csv 파일 윗줄 공백 삭제

yenzip committed:
chore: VoucherException 삭제

yenzip committed:
refactor: yaml 파일 수정
voucher csv 파일 dev 환경에서 동작하도록 수정

shoeone96 committed:
Merge pull request #7 from yenzip/main

[예림] @Profile 적용

ksy90101 commented on logs/error_2023-10-19.log:
```diff
@@ -0,0 +1,4 @@
+21:39:34.178 [main] ERROR o.p.springorder.console.Console -- errorMessage = 잘못된 입력 값입니다. 100 ~ 10000 사이의 값을 입력해주세요.
```
로그 파일을  git에 올리는게 맞을까요?
이건 Ignore를 하는게 좋을거 같아요

ksy90101 commented on src/main/java/org/programmers/springorder/VoucherApplication.java:
```diff
@@ -0,0 +1,42 @@
+
+            switch (menu) {
+                case EXIT -> {
+                    isRunning = false;
+                    console.printMessage(Message.EXIT_PROGRAM_MESSAGE);
```
종료하는 로직도 appliction이 아니라 controller와 같은 곳에서 처리해야 하지 않을까요?

ksy90101 commented on src/test/java/org/programmers/springorder/console/ConsoleTest.java:
```diff
@@ -0,0 +1,78 @@
+        log.info("menuType3 : {}", menuType3.getMenuNum());
+    }
+
+    //TODO: 예외처리로 인해 더 이상 예외가 발생하지 않아서 생기는 test
```
문제가 안된다면 해당 테스트는 삭제하면 좋겠네요.

ksy90101 commented on src/test/java/org/programmers/springorder/console/ConsoleTest.java:
```diff
@@ -0,0 +1,78 @@
+        MenuType menuType3 = console.inputMenu();
+        System.setIn(sysInBackup);
+
+        assertThat(menuType1).isEqualTo(MenuType.EXIT);
+        log.info("menuType1 : {}", menuType1.getMenuNum());
+
+        assertThat(menuType2).isEqualTo(MenuType.CREATE);
+        log.info("menuType2 : {}", menuType2.getMenuNum());
+
+        assertThat(menuType3).isEqualTo(MenuType.LIST);
+        log.info("menuType3 : {}", menuType3.getMenuNum());
```
테스트 코드에서 log를 찍는 이유가 무엇인가요?

ksy90101 commented on src/test/java/org/programmers/springorder/console/ConsoleTest.java:
```diff
@@ -0,0 +1,78 @@
+        InputStream sysInBackup = System.in;
+        System.setIn(new ByteArrayInputStream(simulatedInput.getBytes()));
+        Console console = new Console(); // TODO: 테스트 클래스 필드로 콘솔 초기화하면 사용이 안 됨!
+//        Scanner scanner = new Scanner(System.in);
+
+//        String selectedMenu = scanner.nextLine();
+//        MenuType menuType = MenuType.selectMenu(selectedMenu);
```
주석은 제거하면 좋겠네요.

ksy90101 commented on src/main/java/org/programmers/springorder/console/Console.java:
```diff
@@ -0,0 +1,102 @@
+        } catch (InputMismatchException e) {
+            logger.error("errorMessage = {}", e.getMessage());
+            printMessage(e.getMessage());
+            return inputMenu(); // TODO: 재귀로 호출하는 게 괜찮은 건지 확인 필요
```
재귀가 필요하다면 괜찮습니다.

ksy90101 commented on src/main/java/org/programmers/springorder/console/Console.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    public void showList(List<VoucherResponseDto> allVoucher) {
+        if (allVoucher.size() == 0) {
```
```suggestion
        if (allVoucher.isEmpty()) {
```
이렇게 메서드를 활용해보면 어떨까요?

ksy90101 commented on src/main/java/org/programmers/springorder/customer/repository/CustomerRepository.java:
```diff
@@ -0,0 +1,21 @@
+package org.programmers.springorder.customer.repository;
+
+import org.programmers.springorder.consts.ErrorMessage;
+import org.programmers.springorder.customer.model.Customer;
+import org.programmers.springorder.voucher.model.Voucher;
+import org.programmers.springorder.voucher.model.VoucherType;
+import org.springframework.stereotype.Repository;
+
+import java.io.BufferedReader;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
```
<img width="971" alt="image" src="https://github.com/prgrms-be-devcourse/springboot-basic/assets/53366407/c1d67928-e431-4e94-9390-2b02d6137521">
save option을 이용해 이러한 사용하지 않은 import를 정리리해보세요.

ksy90101 commented on src/main/java/org/programmers/springorder/consts/ErrorMessage.java:
```diff
@@ -0,0 +1,13 @@
+package org.programmers.springorder.consts;
+
+public class ErrorMessage {
```
https://github.com/Meet-Coder-Study/book-effective-java/blob/main/2%EC%9E%A5/4_%EC%9D%B8%EC%8A%A4%ED%84%B4%EC%8A%A4%ED%99%94%EB%A5%BC_%EB%A7%89%EC%9C%BC%EB%A0%A4%EA%B1%B0%EB%93%A0_private_%EC%83%9D%EC%84%B1%EC%9E%90%EB%A5%BC_%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC_%EC%9D%B4%ED%98%B8%EB%B9%88.md

해당글을 한번 참고해보면 어떨까요?

ksy90101 commented on src/main/java/org/programmers/springorder/customer/repository/FileCustomerRepository.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Value(("${customerListFilePath}"))
+    private String filePath;
+    private final Logger logger = LoggerFactory.getLogger(FileVoucherRepository.class);
```
```suggestion
    private static final Logger logger = LoggerFactory.getLogger(FileCustomerRepository.class);
```

1. 상수를 명확히 선언해주시면 좋을거 같아요.
2. 변수보다 상수를 위에 두는게 좋지 않을까요?

ksy90101 commented on src/main/java/org/programmers/springorder/customer/repository/FileCustomerRepository.java:
```diff
@@ -0,0 +1,67 @@
+    @Override
+    public List<Customer> findAllBlackList() {
+        return findAll().stream()
+                .filter(customer -> customer.getCustomerType() == CustomerType.BLACK)
```
custom에게 직접 메시지를 던지면 어떨까요?
그렇다면 getter를 쓰지 않아도 될꺼에요.

ksy90101 commented on src/main/java/org/programmers/springorder/customer/repository/FileCustomerRepository.java:
```diff
@@ -0,0 +1,67 @@
+    public List<Customer> findAllBlackList() {
+        return findAll().stream()
+                .filter(customer -> customer.getCustomerType() == CustomerType.BLACK)
+                .toList();
```
외부에서 리스트를 변경하면 문제가 생기지 않을까요?
따라서 불변을 유지하면 어떨까요?
https://lts0606.tistory.com/133
해당글을 참고해보세요 :)

ksy90101 commented on src/main/java/org/programmers/springorder/voucher/repository/FileVoucherRepository.java:
```diff
@@ -0,0 +1,81 @@
+    @Override
+    public UUID save(Voucher voucher) {
+        try(BufferedWriter bw = new BufferedWriter(new FileWriter(filePath, true))) {
+            String data = String.valueOf(voucher.getVoucherId()) + "," +
+                    String.valueOf(voucher.getDiscountValue()) + "," +
+                    voucher.getVoucherType().name();
```
1. String.valueOf를 사용할 필요가 있을까요?
2. 스트링 연산은 비싸기 때문에 단순 연산보다 더 좋은 방법이 있을거 같습니다.
3. voucher.getVoucherType().name(); 디미터의 법칙을 위반했습니다. 디미터의 법칙이 무엇인지 한번 찾아보시겠어요?

ksy90101 commented on src/test/java/org/programmers/springorder/repository/MemoryVoucherRepositoryTest.java:
```diff
@@ -0,0 +1,96 @@
+        Optional<Voucher> findVoucher = voucherRepository.findById(UUID.randomUUID());
+
+        // then
+        assertThat(findVoucher.isEmpty()).isTrue();
```
```suggestion
        assertThat(findVoucher).isEmpty();
```

ksy90101 commented on src/main/java/org/programmers/springorder/utils/Validation.java:
```diff
@@ -0,0 +1,40 @@
+        return input;
+    }
+
+    public static int validateInt(String input) {
```
해당 메서드는 사용되지 않네요.

ksy90101 commented on src/main/java/org/programmers/springorder/utils/Validation.java:
```diff
@@ -0,0 +1,40 @@
+
+public class Validation {
+    public static String validateString(String input) {
+        if (input == null || input.trim().isEmpty()) {
```
isEmpty()와 isBlnak의 차이를 한번 찾아보시겠어요?
https://rutgo-letsgo.tistory.com/364
얼마전에 제가 공유한 글에도 있는 내용이죠.

ksy90101 commented on src/main/java/org/programmers/springorder/voucher/model/VoucherType.java:
```diff
@@ -0,0 +1,41 @@
+    PERCENT("2", ((beforeDiscount, discount) -> beforeDiscount * (100 - discount) / 100), 3, 50);
+
+    private final String voucherNum;
+    private final BiFunction<Long, Long, Long> expression;
```
LongBinaryOperator도 한번 알아보면 어떨까요?

ksy90101 commented on src/main/java/org/programmers/springorder/voucher/model/VoucherType.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.function.BiFunction;
+
+public enum VoucherType {
+    FIXED("1", ((beforeDiscount, discount) -> beforeDiscount - discount), 100, 10000),
+    PERCENT("2", ((beforeDiscount, discount) -> beforeDiscount * (100 - discount) / 100), 3, 50);
```
이건 1단계 요구사항이 아닌거 같아요?

ksy90101 commented on src/main/java/org/programmers/springorder/voucher/model/VoucherType.java:
```diff
@@ -0,0 +1,41 @@
+    public static VoucherType selectVoucherType(String voucherNum) {
+        return Arrays.stream(VoucherType.values())
+                .filter(voucher -> voucher.voucherNum.equals(voucherNum))
+                .findAny()
```
1. findAny와 findFirst를 비교해보시면 어떨까요?
2. voucher.voucherNum.equals(voucherNum) 이부분을 직접 메시지를 보내서 처리하면 어떨까요?

ksy90101 commented on src/main/java/org/programmers/springorder/voucher/repository/FileVoucherRepository.java:
```diff
@@ -0,0 +1,81 @@
+    @Override
+    public Optional<Voucher> findById(UUID voucherId) {
+        return findAll().stream()
+                .filter(voucher -> voucher.getVoucherId() == voucherId)
```
여기도 메시지를 보내서 처리한다면 getter를 쓰지 않아도 괜찮을거 같습니다.

ksy90101 commented on src/main/java/org/programmers/springorder/voucher/model/Voucher.java:
```diff
@@ -0,0 +1,58 @@
+    private final long discountValue;
+    private final VoucherType voucherType;
+
+    public Voucher(UUID voucherId, long discountValue, VoucherType voucherType) {
+        this.voucherId = voucherId;
+        this.discountValue = discountValue;
+        this.voucherType = voucherType;
+    }
```
정적팩토리 메서드가 있는데 해당 생성자를 만든 이유가 궁금합니다.

ksy90101 commented on src/main/java/org/programmers/springorder/voucher/model/Voucher.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.UUID;
+
+public class Voucher {
+    private final UUID voucherId;
+    private final long discountValue;
+    private final VoucherType voucherType;
```
여기있는 값들중에 validate를 해야 할게 없을까요?

ksy90101 reviewed (changes requested):
고생하셨습니다.
추가적으로 많이 구현해주셨내요.
간단한 피드백을 남겼으니 확인해보세요~

참고로 전체적으로 validate가 필요한 부분이 무엇이 있을지 한번 고민해보시면 좋겠습니다.

shoeone96 committed:
refactor: 16번 수정 사항 isBlank()
input.trim().isEmpty() -> input.isBlank()로 수정

shoeone96 committed:
refactor: 15번 수정 사항 사용하지 않는 메서드
validateInt() -> 사용하지 않는 검증 메서드 삭제

shoeone96 committed:
refactor: 1번 수정 사항 로그 폴더 git
로그 파일 및 폴더 gitignore 추가

shoeone96 committed:
refactor: 3번 수정 사항 사용하지 않는 테스트 삭제
consoleTest 부분 삭제

shoeone96 committed:
refactor: 4번 수정 사항 test 내 불필요한 로그 삭제

shoeone96 committed:
refactor: 7번 수정 사항 list.isEmpty() 사용
list.size() == 0 대신 list.isEmpty() 도입

shoeone96 committed:
refactor: 8번 수정 사항 사용하지 않는 import 문 삭제
Optimize import 도입

shoeone96 committed:
refactor: 9번 수정 사항 유틸 클래스에서 private 생성자를 이용
인스턴스를 생성하지 않는 클래스에 private 생성자 사용

shoeone96 committed:
refactor: 10번 수정 사항 클래스 내 상수 선언 위치 변경
항상 변수 위에 선언하는 것으로 고정

shoeone96 committed:
refactor: 11번 수정 사항 디미터 1
customerBlackList 여부 검증 메서드로 filtering

shoeone96 committed:
refactor: 12번 수정 사항 list 불변 객체 변경
repository 에서 반환하는 객체를 불변객체로 만들기
단 stream의 toList()메서드로 return 하여 불변객체로 취급되기 때문에 따로 처리해주지 않음

shoeone96 committed:
refactor: 13번 수정 사항 save() 메서드 디미터 법칙 적용
voucher 저장 시 사용하는 voucher 정보를 Voucher 클래스 내부에서 제공

shoeone96 committed:
refactor: 14번 수정 사항 isEmpty() 테스트 메서드 사용

shoeone96 committed:
refactor: 17번 수정 사항 LongBinaryOperator 도입
BiFunction -> LongBinaryOperator 변경,
reference type -> primitive type 으로 인수를 받아 사용 가능

shoeone96 committed:
refactor: 19번 수정 사항 findFirst() 도입

shoeone96 committed:
refactor: 20번 수정 사항 내부 검증 메서드 도입 및 if 문 refactoring

shoeone96 committed:
refactor: 21번 수정 사항 Voucher 정적 팩토리 메서드 추가
바우처 생성 및 조회를 위한 정적 팩토리 메서드 추가

yenzip committed:
refactor: 사용하지 않는 변수 제거
age, email 제거

yenzip committed:
test: Customer 정적 팩토리 메서드 사용

yenzip committed:
build: JDBC와 H2 dependency 추가

yenzip committed:
feat: JdbcCustomerRepository 생성
전체 회원 조회 메서드 추가

yenzip committed:
test: JdbcCustomerRepository 전체 회원 조회 테스트
전체 회원 조회 성공 테스트 수행

shoeone96 committed:
Merge pull request #8 from yenzip/main

[예림] JdbcCustomerRepository

shoeone96 committed:
Auto stash before merge of "main" and "origin/main"

shoeone96 committed:
build: mysql 설정 추가

shoeone96 committed:
chore: gitignore 설정 추가

shoeone96 committed:
feat: 회원을 저장하는 기능 추가

shoeone96 committed:
feat: 회원을 저장하는 jdbc 로직 추가

shoeone96 committed:
test: 회원 저장 및 조회 테스트 수행

shoeone96 committed:
feat: 회원 저장 테스트 확인을 위해 동등성을 검증하는 equals, hashcode 메서드 추가
